mkosi/kmod.py

```diff
@@ -380,25 +380,34 @@ def gen_required_kernel_modules(
         # /usr/lib/firmware/nvidia/tu106/gsp
         # /usr/lib/firmware/nvidia/tu117/gsp
 
-        for p in reversed(path_and_parents_below(context.root / fw, context.root / firmwared)):
-            # Make sure only the last component of the path we're working is a symlink. We don't want to
-            # yield any paths with multiple components in them that are symlinks as these paths are passed to
-            # cpio and it interprets intermediary path components that are symlinks as regular directories.
-            # Additionally, intermediary path components that are symlinks also mess up the path
-            # deduplication we want to get by putting all the firmware to add in a set as the paths will hash
-            # differently even though they're the same paths after resolution.
-            p = p.parent.resolve(strict=True).joinpath(p.name)
+        todo = list(reversed(fw.parts))
+        current = context.root
+        while todo:
+            part = todo.pop()
+            if part == ".":
```

Comment:
`fw` is guaranteed to only contain relative paths, since `filter_kernel_modules` returns relative paths and `filter_firmware` keeps them that way, but adding
```suggestion
            if part == "." or part == "/":
```
might be a good idea anyway, since any stray `/` would otherwise clobber `current` to the real root otherwise. This also makes this easier to read, since I had to go back to the above functions to check they don't return absolute paths.

Handling this here, could also allow rewriting the check below as
```python
current = context.root if target.is_absolute() else current.parent
```